sklearn/metrics/_pairwise_distances_reduction/_argkmin_lrd.pyx.tp

```diff
@@ -0,0 +1,129 @@
+from cython cimport floating, integral
+from cython.parallel cimport prange
+from libc.math cimport fmax
+from libc.stdlib cimport free
+
+from ...utils._typedefs cimport intp_t, float64_t
+from ...utils.fixes import threadpool_limits
+
+import numpy as np
+
+
+cdef float64_t EPS = 1e-10
+
+{{for name_suffix in ["32", "64"]}}
+
+from ._argkmin cimport ArgKmin{{name_suffix}}
+from ._datasets_pair cimport DatasetsPair{{name_suffix}}
+
+cdef class ArgKminLRD{{name_suffix}}(ArgKmin{{name_suffix}}):
+    """
+    {{name_suffix}}bit implementation of ArgKminLRD
+    """
+    cdef:
+        float64_t[::1] lrd
+        float64_t[:, ::1] actual_distances
+        intp_t[:, ::1] actual_indices
+        intp_t factor
+
+    @classmethod
+    def compute(
+        cls,
+        X,
+        Y,
+        intp_t k,
+        str metric="euclidean",
+        chunk_size=None,
+        dict metric_kwargs=None,
+        str strategy=None,
+        bint train=False,
+        bint return_distance=False,
+    ):
+        # Use a generic implementation that handles most scipy
+        # metrics by computing the distances between 2 vectors at a time.
+        pda = ArgKminLRD{{name_suffix}}(
+            datasets_pair=DatasetsPair{{name_suffix}}.get_for(X, Y, metric, metric_kwargs),
+            k=k,
+            chunk_size=chunk_size,
+            strategy=strategy,
+            train=train,
+        )
+
+        # Limit the number of threads in second level of nested parallelism for BLAS
+        # to avoid threads over-subscription (in GEMM for instance).
+        with threadpool_limits(limits=1, user_api="blas"):
+            if pda.execute_in_parallel_on_Y:
+                pda._parallel_on_Y()
+            else:
+                pda._parallel_on_X()
+
+        return pda._finalize_results(return_distance)
+
+    def __init__(
+        self,
+        DatasetsPair{{name_suffix}} datasets_pair,
+        chunk_size=None,
+        strategy=None,
+        intp_t k=1,
+        bint train=False,
+    ):
+        super().__init__(
+            datasets_pair=datasets_pair,
+            chunk_size=chunk_size,
+            strategy=strategy,
+            k=k,
+        )
+
+        self.lrd = np.zeros(self.n_samples_X, dtype=np.float64)
+        self.factor = 1 if train else 0
```

Comment:
Forgive me, I am a bit unfamiliar with the implementation details of the algorithm -- what is the purpose of this factor?

Reply:
Actually I initially thought about using the same backend for both the fit and the score_samples method and there are some differences among. This factor handles that because when we consider the fit method we actually have total samples = num_samples + 1. However then I realized that it is not that easy to use this same backend for the score_samples method because in that we also need the distances that were computed during fit. I think since this PR focuses on the fit method for I think this is just an additional detail.